src/operator/contrib/csr_neighborhood_sample-inl.h

```diff
@@ -0,0 +1,281 @@
+      }
+    }
+  }
+}
+
+static void CSRNeighborSampleComputeExCPU(const nnvm::NodeAttrs& attrs,
+                                          const OpContext& ctx,
+                                          const std::vector<NDArray>& inputs,
+                                          const std::vector<OpReqType>& req,
+                                          const std::vector<NDArray>& outputs) {
+  CHECK_EQ(inputs.size(), 3U);
+  CHECK_EQ(outputs.size(), 1U);
+
+  const NeighborSampleParam& params = 
+    nnvm::get<NeighborSampleParam>(attrs.parsed);
+
+  // set seed for random sampling
+  srand(time(nullptr));
+
+  dgl_id_t num_hops = params.num_hops;
+  dgl_id_t num_neighbor = params.num_neighbor;
+  dgl_id_t max_num_vertices = params.max_num_vertices;
+
+  size_t vertices_num = inputs[1].data().Size();
+  size_t egde_num = inputs[0].data().Size();
+  size_t seed_num = inputs[2].data().Size();
+
+  CHECK_GE(max_num_vertices, seed_num);
+
+  const dgl_id_t* val_list = inputs[0].data().dptr<dgl_id_t>();
+  const dgl_id_t* col_list = inputs[0].aux_data(1).dptr<dgl_id_t>();
+  const dgl_id_t* indptr = inputs[0].aux_data(0).dptr<dgl_id_t>();
+  const dgl_id_t* seed = inputs[2].data().dptr<dgl_id_t>();
+  const float* prob_array = inputs[1].data().dptr<float>();
+
+  size_t row_len = inputs[0].aux_data(0).Size() - 1;
+  CHECK_EQ(row_len, vertices_num);
+
+  dgl_id_t* out = outputs[0].data().dptr<dgl_id_t>();
+
+  // Get the mapping between edge_id and row_id
+  dgl_id_t idx = 0;
+  std::unordered_map<dgl_id_t, dgl_id_t> edge_mp;
+  for (size_t i = 1; i < vertices_num+1; ++i) {
+    size_t edge_in_a_row = indptr[i] - indptr[i-1];
+    for (size_t j = 0; j < edge_in_a_row; ++j) {
+      edge_mp[val_list[idx++]] = i-1;
+    }
+  }
+
+  // Get the mapping between col_id (src) and row_id_list (dst)
+  std::vector<std::vector<dgl_id_t> > col_row_list(vertices_num);
+  for (size_t i = 0; i < egde_num; ++i) {
+    dgl_id_t row_id = edge_mp[val_list[i]];
+    dgl_id_t col_id = col_list[i];
+    col_row_list[col_id].push_back(row_id);
+  }
+
+  // BFS traverse the graph and sample vertices
+  std::vector<dgl_id_t> sub_ver_mp(vertices_num, 0);
+  std::queue<ver_node> node_queue;
+  dgl_id_t vertices_count = 0;
+  for (size_t i = 0; i < seed_num; ++i) {
+    // add seed vertices
+    ver_node node;
+    node.vertex_id = seed[i];
+    node.level = 0;
+    node_queue.push(node);
+    // use 1 as flag
+    sub_ver_mp[node.vertex_id] = 1;
+    vertices_count++;
+  }
+
+  std::vector<dgl_id_t> sampled_vec;
+  while (!node_queue.empty() && 
+         vertices_count < max_num_vertices) {
+    ver_node& cur_node = node_queue.front();
+    node_queue.pop();
+    if (cur_node.level < num_hops) {
+      dgl_id_t src_id = cur_node.vertex_id;
+      sampled_vec.clear();
+      GetSample(col_row_list[src_id], // ver_list
+                prob_array,      // probability
+                num_neighbor,    // max_num_neighbor
+                sampled_vec);    // output
```

Comment:
actually, there is an easier case we should support first, i.e., neighbors are sampled uniformly. I think we should distinguish the easier case with the general case for the sake of performance. However we implement the general case, it's going to be much more expensive. We care about the performance of subgraph sampling a lot.

Reply:
@zheng-da Do we need to change the API or we can just give a uniform probability to this operator?

Reply:
We don't need to change the interface of the operator. One option is to give an NDArray with one element.